### ISSUE TITLE
Themes

### ISSUE THREAD

**Issue by HipyCas** [The-Noah/lsfp#56]
What I think may be a great idea for lsfp came to my mind recently, and that idea are themes. By themes I mean custom color sets, maybe some defaults that you can choose from, and also user designed ones (for example on a file). Keep in mind, this is a concept, to be done in a future and to keep developing, it is not a complete concept by any means.

So, the themes would be like the actual `FILE_EXTENSION_COLORS` constant. By default, the colors that we have registered there would be used, but the user would have the option to decide among other themes, built-in or custom. The theme selection could be done either by a `--theme` flag or a `theme` option in your `.lsfprc`, which would take a string or path. The string passed should be the name of one of the built-in themes like `noir`, `mono`, `solarized`, etc. (just some ideas). On the other hand, if a path is passed, lsfp would locate the file and read the theme from it. Summing up, to get a different theme for that print, you would do `lsfp --theme <builtin_theme_name | path_to_theme_file>`, and to make it persistent, a `theme=<builtin_theme_name | path_to_theme_file>` setting in your `.lsfprc`.

For the built-in themes, rather than hard-coding them, which could work, maybe we could distribute them as an additional zip file. This file would contain a bunch of theme files, which then the user would unzip for example in `~/.lsfp-themes` directory, which lsfp would read if a named theme is passed to it, by finding the file whose name is the same as the passed. This would also allow users to define named custom named themes, just by creating the theme file in that directory. The main benefit of making a separate zip for it rather than hard-coding it is that the themes don't contribute to the binary size and people who are fine with the default styling don't have to worry about themes.

For the theme definition files, we can look at some already existing syntax and maybe simplify it, which would make it easier for anyone to use, or we could make an easy to parse custom syntax. Just came up with this:

```
--- javascript: The text after the first dash (can be more than one) does not matter, the parser will start to parse again after the newline
extensions=["js", "mjs"]
color=(0, 0, 0)
icon=e54f
- typescript: Users could use this header to write the file type or make some comments, nothing will get parsed!
extensions=["ts"]
color=(0, 0, 255)
icon=e55f
---------- <other_language>
...
```

The syntax would be based in a dash (`-`) to separate languages, and anything written after the dash until the newline would be ignored, that way leaving room for the user to type the language name or add some comments. Then, items would be a `<key>=<value>` pair, which we could borrow from the already existing config parser. For extensions, a classic square bracket array shouldn't be hard to parse I think, we could maybe even drop the quotes around the values for even more simplicity (another idea is dropping everything except the comma, which could also work or even be easier to parse). The color is a simple rgb set in the form `(r, g, b)`, which shouldn't be something complicated to parse either. The icon (yet to introduce, #47) is two hexadecimal values, adding up to 4 characters, which would definitely be something easy to parse, is just taking the next 4 characters or maybe to the new line. **This is only a concept, it may then be too complicated or even impossible to parse.**.

I will try to finish the icons now, and when I'm done with that I will try to look into this if you think it is a good idea, and as I go I will try to show you some snapshots of what the themes are looking like.

**Comment by HipyCas** [comment]
I have recently been working on this feature, now that icons has already been pushed. I was getting a bit confused with the many code that was around me, lifetimes and other aspects, and therefore I decided to work on the parser in a new repository. So I have concluded the basic aspects of the parser, being it fully functional, although there are some aspects left to polish. You can find the code at [HipyCas/theme-parser](https://github.com/HipyCas/theme-parser). It would be great if you could take a look and give me your opinion! You can also try to play with the code and try to break it some way!

**Comment by The-Noah** [comment]
It looks good.

I'm thinking using `#` for comments makes more sense, as that's used in many languages and config files.

**Comment by HipyCas** [comment]
After much work on the parser, I have come with a highly satisfactory experience. I won't rewrite every explanation about the theme parser implementation here, instead got to the [formal specification in HipyCas/theme-parser](https://github.com/HipyCas/theme-parser/blob/master/SPEC.md) (you could also look at the README, but it is **highly** outdated, so I recommend the specification over the README. I encourage you to clone the repo and test the program, with different theme files (try to break it too!). If everything looks good to you, I'll copy the code to the mainstream codebase and make the necessary modifications to other `lsfp` related code.

For what I called _named themes_, which are basically just theme files that you can invoke by a name instead of a path, I decided to make them live in `~/.lsfp-themes/`. This works by reading the files in the named folder, and matching the passed _named theme_ with the file names, e.g. `lsfp --theme noir` will result in loading `~/.lsfp-themes/noir` and `lsfp --theme solarized.dark` will result in loading `~/.lsfp-themes/solarized.dark`. I took the whole file and decided not to strip any extension simply because I hadn't thought of any extension and there is currently (and I don't think there will ever be) any support for this theme file. What do you think of the idea?

I had to come up with some creative ideas for parsing the value out of the `--theme` option, but it is working. The only real problem I found was getting a value from `.lsfprc`, which as of now is only used for boolean values. I will do some work later on on that, but the fist version of themes will probably go out without support for configuration in `.lsfprc`.

I'm waiting for your opinion on the parser, and for the approval. Feel free to ask me about it and tell me any problems you may encounter with it.

**Comment by The-Noah** [comment]
Looks great!

I don't think there's any need for a file extension.
